benchmark/multithread.cpp

```diff
@@ -0,0 +1,121 @@
+/*
+ *
+ * Copyright (C) 2024 Intel Corporation
+ *
+ * Under the Apache License v2.0 with LLVM Exceptions. See LICENSE.TXT.
+ * SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+ *
+ */
+
+#include "multithread.hpp"
+
+#include <umf/memory_pool.h>
+#include <umf/pools/pool_disjoint.h>
+#include <umf/pools/pool_jemalloc.h>
+#include <umf/pools/pool_scalable.h>
+#include <umf/providers/provider_os_memory.h>
+
+#include <iostream>
+#include <memory>
+#include <numeric>
+
+static constexpr size_t N_REPEATS = 5;
+static constexpr size_t N_ITERATIONS = 50000;
+static constexpr size_t N_THREADS = 20;
+static constexpr size_t ALLOC_SIZE = 64;
+
+using poolCreateExtParams = std::tuple<umf_memory_pool_ops_t *, void *,
+                                       umf_memory_provider_ops_t *, void *>;
+
+static auto poolCreateExtUnique(poolCreateExtParams params) {
+    umf_memory_pool_handle_t hPool;
+    auto [pool_ops, pool_params, provider_ops, provider_params] = params;
+
+    umf_memory_provider_handle_t provider = nullptr;
+    auto ret =
+        umfMemoryProviderCreate(provider_ops, provider_params, &provider);
+    if (ret != UMF_RESULT_SUCCESS) {
+        std::cerr << "provider create failed" << std::endl;
+        abort();
+    }
+
+    ret = umfPoolCreate(pool_ops, provider, pool_params,
+                        UMF_POOL_CREATE_FLAG_OWN_PROVIDER, &hPool);
+    if (ret != UMF_RESULT_SUCCESS) {
+        std::cerr << "pool create failed" << std::endl;
+        abort();
+    }
+
+    return std::shared_ptr<umf_memory_pool_t>(hPool, &umfPoolDestroy);
+}
+
+static void mt_alloc_free(poolCreateExtParams params) {
+    auto pool = poolCreateExtUnique(params);
+
+    std::vector<void *> allocs[N_THREADS];
+    size_t numFailures[N_THREADS] = {};
+    for (auto &v : allocs) {
+        v.reserve(N_ITERATIONS);
+    }
+
+    auto values = umf_bench::measure<std::chrono::milliseconds>(
+        N_REPEATS, N_THREADS, [&, pool = pool.get()](auto thread_id) {
+            for (int i = 0; i < N_ITERATIONS; i++) {
+                allocs[thread_id].push_back(umfPoolMalloc(pool, ALLOC_SIZE));
+                if (!allocs[thread_id].back()) {
+                    numFailures[thread_id]++;
+                }
+            }
+
+            for (int i = 0; i < N_ITERATIONS; i++) {
+                umfPoolFree(pool, allocs[thread_id][i]);
+            }
+
+            // clear the vector as this function might be called multiple times
+            allocs[thread_id].clear();
+        });
+
+    std::cout << "mean: " << umf_bench::mean(values)
+              << " [ms] std_dev: " << umf_bench::std_dev(values) << " [ms]"
+              << " (total alloc failures: "
+              << std::accumulate(numFailures, numFailures + N_THREADS, 0)
+              << " out of " << N_ITERATIONS * N_REPEATS * N_THREADS << ")"
+              << std::endl;
+}
+
+int main() {
+#if defined(UMF_BUILD_OS_MEMORY_PROVIDER)
```

Comment:
You can make this a requirement to create an MT Benchmark target in CMake and remove these checks.
 If `UMF_BUILD_OS_MEMORY_PROVIDER` is not set, this binary does nothing except printing some skipping messages.

Reply:
I'm actually planning to add some other cases soon that won't require OS provider.